pkg/sql/colexec/index_table.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package colexec
+
+import (
+	"context"
+	"github.com/matrixorigin/matrixone/pkg/catalog"
+	"github.com/matrixorigin/matrixone/pkg/container/batch"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/vm/engine"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+func BuildRowidBatch(ctx context.Context, rel engine.Relation, b *batch.Batch, proc *process.Process) (*batch.Batch, error) {
+	v := b.Vecs[0]
+	var rds []engine.Reader
+	ret, err := rel.Ranges(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	rds, _ = rel.NewReader(ctx, 1, nil, ret)
+
+	readBat, err := rds[0].Read([]string{catalog.Row_ID, b.Attrs[0]}, nil, proc.Mp())
+	if err != nil {
+		return nil, err
+	}
+	r := vector.MustTCols[types.Rowid](readBat.Vecs[0])
+	newVec := vector.New(readBat.Vecs[0].Typ)
+	switch v.Typ.Oid {
+	case types.T_bool:
+		s := vector.MustTCols[bool](v)
+		ns := vector.MustTCols[bool](readBat.Vecs[1])
+		for _, b := range s {
```

Comment:
What is the size of s and ns?   1M * 1M, this loop will take forever.

Reply:
the size of n and ns is the delete batch size and the table row size, I will optimize the code later